pennylane_sf/tf.py

```diff
@@ -0,0 +1,275 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Strawberry Fields TF backend for PennyLane.
+"""
+from collections import OrderedDict
+from collections.abc import Sequence  # pylint: disable=no-name-in-module
+import uuid
+
+import numpy as np
+import tensorflow as tf
+
+import strawberryfields as sf
+from strawberryfields.backends.tfbackend.states import FockStateTF
+
+# import state preparations
+from strawberryfields.ops import (
+    # Catstate,
+    Coherent,
+    DensityMatrix,
+    DisplacedSqueezed,
+    Fock,
+    Ket,
+    Squeezed,
+    Thermal,
+    Gaussian,
+)
+
+# import gates
+from strawberryfields.ops import (
+    BSgate,
+    CKgate,
+    CXgate,
+    CZgate,
+    Dgate,
+    Kgate,
+    Pgate,
+    Rgate,
+    S2gate,
+    Sgate,
+    Vgate,
+    Interferometer,
+)
+
+from pennylane.wires import Wires
+
+from .expectations import mean_photon, number_expectation, homodyne, poly_xp
+from .simulator import StrawberryFieldsSimulator
+
+
+def identity(state, device_wires, params):
+    """Computes the expectation value of ``qml.Identity``
+    observable in Strawberry Fields, corresponding to the trace.
+
+    Args:
+        state (strawberryfields.backends.states.BaseState): the quantum state
+        device_wires (Wires): the measured modes
+        params (Sequence): sequence of parameters (not used)
+
+    Returns:
+        float, float: trace and its variance
+    """
+    # pylint: disable=unused-argument
+    N = state.num_modes
+    D = state.cutoff_dim
+
+    if N == len(device_wires):
+        # trace of the entire system
+        tr = state.trace()
+        return tr, tr - tr ** 2
+
+    # get the reduced density matrix
+    N = len(device_wires)
+    dm = state.reduced_dm(modes=device_wires.tolist())
+
+    # construct the standard 2D density matrix, and take the trace
+    new_ax = np.arange(2 * N).reshape([N, 2]).T.flatten()
+    tr = tf.math.real(tf.linalg.trace(tf.reshape(tf.transpose(dm, new_ax), [D ** N, D ** N])))
+
+    return tr, tr - tr ** 2
+
+
+def fock_state(state, device_wires, params):
+    """Computes the expectation value of the ``qml.FockStateProjector``
+    observable in Strawberry Fields.
+
+    Args:
+        state (strawberryfields.backends.states.BaseState): the quantum state
+        device_wires (Wires): the measured mode
+        params (Sequence): sequence of parameters
+
+    Returns:
+        float, float: Fock state probability and its variance
+    """
+    # pylint: disable=unused-argument
+    n = params[0]
+    N = state.num_modes
+
+    if N == len(device_wires):
+        # expectation value of the entire system
+        ex = state.fock_prob(n)
+        return ex, ex - ex ** 2
+
+    dm = state.reduced_dm(modes=device_wires.tolist())
+    ex = tf.math.real(dm[tuple([n[i // 2] for i in range(len(n) * 2)])])
+
+    var = ex - ex ** 2
+    return ex, var

```

Comment:
> Is it because the state isn't a BaseFockState, and thus fock_state doesn't return the correct values? 

No, actually the `FockStateTF` class that the `tf` backend returns is a subclass of `BaseFockState`, so `isinstance(state, BaseFockState)` would still work!

The issue is solely the `.real` in the other function; we need to replace all NumPy operations with TensorFlow operations (`tf.math.real`) to preserve differentiability, otherwise the gradient will just be `None`.

Reply:
Oh, that makes more sense. I missed that change. Thanks! 😄